rust/datafusion/src/execution/context.rs

```diff
@@ -678,20 +695,31 @@ impl ExecutionContextState {
             .resolve(&self.config.default_catalog, &self.config.default_schema)
     }
 
+    fn catalog_for_ref<'a>(
+        &'a self,
+        resolved_ref: &ResolvedTableReference<'a>,
+    ) -> Result<Arc<dyn CatalogProvider>> {
+        if self.config.information_schema {
+            CatalogListWithInformationSchema::new(self.catalog_list.clone())
```

Comment:
Not a massive issue as this isn't the critical path, but would it be worth doing this wrapping on catalog registration so we avoid query-side allocs here? I wonder if it's possible to remove the `CatalogListWithInformationSchema` entirely and just do this inside `register_catalog` instead. So if the info schema is enabled, wrap the catalog up in a `CatalogWithInformationSchema` before registering it with the list, or otherwise, register the plain catalog.

Reply:
Since the `Config` can't be changed after the `ExecutionContext` is created I think this would be a good idea. I will give it a try.

Reply:
in https://github.com/apache/arrow/pull/9818/commits/5151f31eecfbcd0d5892837f951bb0e5ea8473c0